File: src/cpp/Source/Chunks/Rdef/ConstantBuffer.cpp
#include "PCH.h"
#include "ConstantBuffer.h"

using namespace std;
using namespace SlimShader;

ConstantBuffer ConstantBuffer::Parse(const BytecodeReader& reader,
									 BytecodeReader& constantBufferReader,
									 const ShaderVersion& target)
{
	auto nameOffset = constantBufferReader.ReadUInt32();
	auto nameReader = reader.CopyAtOffset(nameOffset);

	auto variableCount = constantBufferReader.ReadUInt32();
	auto variableOffset = constantBufferReader.ReadUInt32();

	ConstantBuffer result;

	result._name = nameReader.ReadString();
	result._size = constantBufferReader.ReadUInt32();
	result._flags = static_cast<ConstantBufferFlags>(constantBufferReader.ReadUInt32());
	result._bufferType = static_cast<ConstantBufferType>(constantBufferReader.ReadUInt32());

	auto variableReader = reader.CopyAtOffset(variableOffset);
	for (uint32_t i = 0; i < variableCount; i++)
		result._variables.push_back(ShaderVariable::Parse(reader, variableReader, target, i == 0));

	return result;
}

const string& ConstantBuffer::GetName() const { return _name; }
ConstantBufferType ConstantBuffer::GetBufferType() const { return _bufferType; }
const vector<ShaderVariable>& ConstantBuffer::GetVariables() const { return _variables; }
uint32_t ConstantBuffer::GetSize() const { return _size; }
ConstantBufferFlags ConstantBuffer::GetFlags() const { return _flags; }

ostream& SlimShader::operator<<(ostream& out, const ConstantBuffer& value)
{
	out << "// " << ToString(value._bufferType) << " " << value._name << endl;
	out << "// {" << endl;

	for (auto& variable : value._variables)
		out << variable;

	out << "//" << endl;
	out << "// }" << endl;
	out << "//" << endl;

	return out;
}

File: src/SlimShader/Chunks/Rdef/ConstantBuffer.cs
using System.Collections.Generic;
using System.Text;
using SlimShader.Chunks.Common;
using SlimShader.Chunks.Shex;
using SlimShader.Util;

namespace SlimShader.Chunks.Rdef
{
	/// <summary>
	/// Describes a shader constant-buffer.
	/// Based on D3D11_SHADER_BUFFER_DESC.
	/// </summary>
	public class ConstantBuffer
	{
		/// <summary>
		/// The name of the buffer.
		/// </summary>
		public string Name { get; private set; }

		/// <summary>
		/// A <see cref="ConstantBufferType" />-typed value that indicates the intended use of the constant data.
		/// </summary>
		public ConstantBufferType BufferType { get; private set; }

		public List<ShaderVariable> Variables { get; private set; }

		/// <summary>
		/// Buffer size (in bytes).
		/// </summary>
		public uint Size { get; private set; }

		/// <summary>
		/// A combination of <see cref="ConstantBufferFlags" />-typed values that are combined by using a bitwise OR 
		/// operation. The resulting value specifies properties for the shader constant-buffer.
		/// </summary>
		public ConstantBufferFlags Flags { get; private set; }

		public ConstantBuffer()
		{
			Variables = new List<ShaderVariable>();
		}

		public static ConstantBuffer Parse(
			BytecodeReader reader, BytecodeReader constantBufferReader,
			ShaderVersion target)
		{
			uint nameOffset = constantBufferReader.ReadUInt32();
			var nameReader = reader.CopyAtOffset((int) nameOffset);

			uint variableCount = constantBufferReader.ReadUInt32();
			uint variableOffset = constantBufferReader.ReadUInt32();

			var result = new ConstantBuffer
			{
				Name = nameReader.ReadString()
			};

			var variableReader = reader.CopyAtOffset((int) variableOffset);
			for (int i = 0; i < variableCount; i++)
				result.Variables.Add(ShaderVariable.Parse(reader, variableReader, target, i == 0));

			result.Size = constantBufferReader.ReadUInt32();
			result.Flags = (ConstantBufferFlags) constantBufferReader.ReadUInt32();
			result.BufferType = (ConstantBufferType) constantBufferReader.ReadUInt32();

			return result;
		}

		public override string ToString()
		{
			var sb = new StringBuilder();
			sb.AppendLine(string.Format("// {0} {1}", BufferType.GetDescription(), Name));
			sb.AppendLine("// {");

			foreach (var variable in Variables)
				sb.Append(variable);

			sb.AppendLine("//");
			sb.AppendLine("// }");
			sb.AppendLine("//");

			return sb.ToString();
		}
	}
}